cylc/flow/parsec/fileparse.py

```diff
@@ -239,27 +268,63 @@ def read_and_proc(fpath, template_vars=None, viewcfg=None, asedit=False):
         flines = inline(
             flines, fdir, fpath, False, viewcfg=viewcfg, for_edit=asedit)
 
+    template_vars['CYLC_VERSION'] = __version__
+
+    # Push template_vars into extra_vars so that duplicates come from
+    # template_vars.
+    if extra_vars['templating_detected'] is not None:
+        extra_vars['template_variables'].update(template_vars)
+        template_vars = extra_vars['template_variables']
+
     # process with EmPy
     if do_empy:
+        if (
+            extra_vars['templating_detected'] == 'empy' and
+            not re.match(r'^#![Ee]m[Pp]y\s*', flines[0])
+        ):
+            if not re.match(r'^#!', flines[0]):
+                flines.insert(0, '#!empy')
+            else:
+                raise FileParseError(
+                    "Your file has shebang line for "
+                    f"{extra_vars['templating_detected']}"
+                    f" but {flines[0]} configured in flow.cylc."
+                )
         if flines and re.match(r'^#![Ee]m[Pp]y\s*', flines[0]):
             LOG.debug('Processing with EmPy')
             try:
                 from cylc.flow.parsec.empysupport import empyprocess
             except (ImportError, ModuleNotFoundError):
                 raise ParsecError('EmPy Python package must be installed '
                                   'to process file: ' + fpath)
-            flines = empyprocess(flines, fdir, template_vars)
+            flines = empyprocess(
+                flines, fdir, template_vars
+            )
 
     # process with Jinja2
     if do_jinja2:
+        if (
+            extra_vars['templating_detected'] == 'jinja2' and
+            not re.match(r'^#![jJ]inja2\s*', flines[0])
+        ):
+            if not re.match(r'^#!', flines[0]):
+                flines.insert(0, '#!jinja2')
+            else:
+                raise FileParseError(
+                    "Your file has shebang line for "
+                    f"{extra_vars['templating_detected']}"
+                    f" but {flines[0]} configured in flow.cylc."
+                )
         if flines and re.match(r'^#![jJ]inja2\s*', flines[0]):
             LOG.debug('Processing with Jinja2')
             try:
                 from cylc.flow.parsec.jinja2support import jinja2process
             except (ImportError, ModuleNotFoundError):
                 raise ParsecError('Jinja2 Python package must be installed '
                                   'to process file: ' + fpath)
-            flines = jinja2process(flines, fdir, template_vars)
+            flines = jinja2process(
+                flines, fdir, template_vars
+            )
```

Comment:
The logic here looks a bit suspect. I think that if you have template with `#!jinja2` and plugin with `'templating_detected': 'empy'` you end up passing the plugin variables to the template engine regardless.

It might be worth to eliminate this split between jinja2/empy and simplify the logic like so:

```python
# assume viewcfg['empy'] + viewcfg['jinja2'] replaced by viewcfg['template']
do_templating = viewcfg['template']

template_engines = {
    'jinja2': jinja2process,
    'empy': empyprocess,
}

templating_detected = re.match(r'#!(\w+)\s*', flines[0])
templating_detected = templating_detected and templating_detected.group(1).lower()

for entry_point in pkg_resources.iter_entry_points('cylc.pre_configure'):
    ...  # detect collisions and update extra_vars

if extra_vars.get('templating_detected'):
    if not templating_detected:
        templating_detected = extra_vars['templating_detected']
        flines.insert(0, templating_detected)
    elif extra_vars['templating_detected'] != templating_detected:
        raise FileParseError(
            'Mismatch between parsed and configured template processor: %s != %s'
            % (templating_detected, extra_vars['templating_detected'])
        )

template_vars = {**extra_vars['template_variables'], **template_vars}

if do_templating and templating_detected and flines:
    LOG.debug('Processing with ' + templating_detected)
    flines = template_engines[templating_detected](flines, fdir, template_vars)
```

Reply:
This is very likely a good idea, but it's also mission creep. I'm happy to address the suggested re-structure but I'd like to put this PR to bed.

Reply:
@TomekTrzeciak  - If I've understood you correctly the new functional test `08-template-engine-conflict.t` should demonstrate that if you set a shebang in your `flow.cylc` file and the plugin returns a `templating_detected` that is different you will get a validation failure.